ENH number counts integrals

beckermr opened the pull request:
This PR has number counts integrals

damonge commented on pyccl/halos/halo_model.py:
```diff
@@ -137,6 +140,69 @@ def profile_norm(self, cosmo, a, prof):
+        # now do m intergrals in a loop
+        mint = np.zeros_like(a)
+        sint = np.zeros_like(a)
+        for i, _a in enumerate(a):
```
If this is too slow we may want to write a quick 2D integrator that openMPs over a. This wouldn't be too difficult, since we already have a spline integrator at the C level. I had to do something similar for the fftlog wrapper.

beckermr replied:
this seems fast enough for now

damonge commented on pyccl/halos/halo_model.py:
```diff
@@ -137,6 +140,69 @@ def profile_norm(self, cosmo, a, prof):
+        Args:
+            cosmo (:class:`~pyccl.core.Cosmology`): a Cosmology object.
+            sel (callable): function of mass and scale factor that returns the
+                possibly unnormalized selection function
```
There should be some information on what the expected dimensions of the inputs and outputs of this function are

damonge commented on pyccl/halos/halo_model.py:
```diff
@@ -137,6 +140,69 @@ def profile_norm(self, cosmo, a, prof):
+                dvda[i] * self.mf[..., :] * _selm[..., :] * dmdlog10m,
+                self._lmass
+            )
+            sint[i] = self._integrator(_selm[..., :] * dmdlog10m, self._lmass)
```
I am not convinced that the selection function is something that should be normalized, or something that has units of M^-1 (what I said above about it being a probability distribution). You can't have two things that are per unit mass, integrate over mass once and then get something that does not have units of M^-1 (i.e. counts).

damonge commented on pyccl/halos/halo_model.py:
```diff
@@ -137,6 +140,69 @@ def profile_norm(self, cosmo, a, prof):
+
+        # now do scale factor integrals and normalize
+        mtot = self._integrator(mint, a)
+        stot = self._integrator(sint, a)
```
See above

damonge commented on pyccl/halos/halo_model.py:
```diff
@@ -137,6 +140,69 @@ def profile_norm(self, cosmo, a, prof):
+        if np.allclose(stot, 0):
+            return 0
+        else:
+            return mtot / stot
```
And see above

beckermr commented on pyccl/halos/halo_model.py:
```diff
@@ -137,6 +140,69 @@ def profile_norm(self, cosmo, a, prof):
+        da = background.angular_diameter_distance(cosmo, a)
+        ez = background.h_over_h0(cosmo, a)
+        dh = physical_constants.CLIGHT_HMPC / cosmo['h']
+        dvdz = dh * (1.0 + z)**2 * da**2 / ez
```
fixed

beckermr commented on pyccl/halos/halo_model.py:
```diff
@@ -137,6 +140,69 @@ def profile_norm(self, cosmo, a, prof):
+
+        # this is used to convert quantities per dm to per dlog10m (e.g.
+        # the selection function )
+        dmdlog10m = self._mass * np.log(10)
```
fixed

beckermr commented:
Thanks @damonge! I fixed a bunch of things and addressed your comments. The code is still too slow and I can only match the numcosmo benchmarks to a percent or so.

beckermr commented:
OK @damonge. This PR is ready for another look. I'd say overall the integrals have an error around a percent or so. The benchmarks agree to a big better than this, but comparing to an adaptive integrator shows a bigger difference.

damonge commented on pyccl/halos/halo_model.py:
```diff
@@ -137,6 +140,67 @@ def profile_norm(self, cosmo, a, prof):
+
+        Args:
+            cosmo (:class:`~pyccl.core.Cosmology`): a Cosmology object.
+            sel (callable): function of mass and scale factor that returns the
```
Is there any situation where the selection function would be cosmology-dependent?

beckermr replied:
It could be, but this entire integral is at fixed cosmology so we can assume that dependence is fixed.

damonge commented on pyccl/halos/halo_model.py:
```diff
@@ -137,6 +140,67 @@ def profile_norm(self, cosmo, a, prof):
+        mint = np.zeros_like(a)
+        for i, _a in enumerate(a):
+            self._get_ingredients(_a, cosmo, False)
+            _selm = np.atleast_2d(sel(self._mass, _a)).T
```
Since you're transposing things here anyway, would it make sense to just ask that `sel` return arrays of shape ``(len(a), len(m))``?

beckermr replied:
I was trying to match the dimensions of other quantities in the code under the assumption that this was useful. Happy to change.

damonge replied:
Ah wait, so are we doing this stupidly in other places too?

damonge replied:
(and by we I mean me...)

beckermr replied:
yes

damonge replied:
Oh, actually, I remember there was a reason for this, to do with what's best to integrate profiles (which is a limiting factor often). OK, nevermind. Let's approve this.

damonge commented:
To clarify: does the adaptive integrator agree with numcosmo better than CCL?
Note that all the halo model integrals before matched the previous implementation of the halo model in CCL, which did use slow adaptive integrators. The mass function integrands are usually smooth enough that you don't need anything too fancy...

beckermr commented:
> To clarify: does the adaptive integrator agree with numcosmo better than CCL?
Note that all the halo model integrals before matched the previous implementation of the halo model in CCL, which did use slow adaptive integrators. The mass function integrands are usually smooth enough that you don't need anything too fancy...

No. Our spline integrator in the HMCalculator class agrees the best.

beckermr commented:
@damonge I left some comments above. If you are ok with those, then let's merge.
